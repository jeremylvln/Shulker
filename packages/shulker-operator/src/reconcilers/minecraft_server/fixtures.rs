use std::collections::BTreeMap;

use http::{Request, Response};
use k8s_openapi::api::core::v1::EnvVar;
use kube::client::Body;
use kube::{core::ObjectMeta, Client};
use lazy_static::lazy_static;
use shulker_crds::{
    resourceref::ResourceRefSpec,
    v1alpha1::{
        minecraft_cluster::MinecraftClusterRef,
        minecraft_server::{
            MinecraftServer, MinecraftServerConfigurationLifecycleStrategy,
            MinecraftServerConfigurationProxyForwardingMode, MinecraftServerConfigurationSpec,
            MinecraftServerPodOverridesSpec, MinecraftServerSpec, MinecraftServerVersion,
            MinecraftServerVersionSpec,
        },
    },
};

lazy_static! {
    pub static ref TEST_SERVER: MinecraftServer = MinecraftServer {
        metadata: ObjectMeta {
            namespace: Some("default".to_string()),
            name: Some("my-server".to_string()),
            ..ObjectMeta::default()
        },
        spec: MinecraftServerSpec {
            cluster_ref: MinecraftClusterRef::new("my-cluster".to_string()),
            tags: vec!["lobby".to_string()],
            version: MinecraftServerVersionSpec {
                channel: MinecraftServerVersion::Paper,
                name: "1.20.1".to_string(),
                ..MinecraftServerVersionSpec::default()
            },
            config: MinecraftServerConfigurationSpec {
                existing_config_map_name: None,
                world: Some(ResourceRefSpec {
                    url: Some("https://example.com/my_world.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }),
                plugins: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_plugin.jar".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                patches: Some(vec![ResourceRefSpec {
                    url: Some("https://example.com/my_patch.tar.gz".to_string()),
                    ..ResourceRefSpec::default()
                }]),
                max_players: 42,
                disable_nether: false,
                disable_end: true,
                server_properties: None,
                proxy_forwarding_mode: MinecraftServerConfigurationProxyForwardingMode::Velocity,
                lifecycle_strategy:
                    MinecraftServerConfigurationLifecycleStrategy::AllocateWhenNotEmpty,
                skip_agent_download: false
            },
            pod_overrides: Some(MinecraftServerPodOverridesSpec {
                image: None,
                env: Some(vec![EnvVar {
                    name: "EXTRA_ENV".to_string(),
                    value: Some("my_value".to_string()),
                    ..EnvVar::default()
                }]),
                resources: None,
                affinity: None,
                node_selector: Some(BTreeMap::from([(
                    "beta.kubernetes.io/os".to_string(),
                    "linux".to_string()
                )])),
                tolerations: None,
                service_account_name: None,
                volume_mounts: None,
                volumes: None,
                ports: None,
                health: None,
            })
        },
        status: None,
    };
}

pub fn create_client_mock() -> Client {
    let (mock_service, _) = tower_test::mock::pair::<Request<Body>, Response<Body>>();
    Client::new(mock_service, "default")
}
