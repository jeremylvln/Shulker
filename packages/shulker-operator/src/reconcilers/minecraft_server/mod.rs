use std::{collections::BTreeMap, sync::Arc, time::Duration};

use futures::StreamExt;
use google_agones_crds::v1::game_server::GameServer;
use k8s_openapi::api::core::v1::ConfigMap;
use kube::{
    api::{DeleteParams, ListParams, PatchParams},
    runtime::{controller::Action, watcher::Config, Controller},
    Api, Client, ResourceExt,
};
use shulker_kube_utils::reconcilers::{builder::reconcile_builder, status::patch_status};
use tracing::*;

use shulker_crds::{
    condition::{ConditionStatus, HasConditions},
    v1alpha1::minecraft_server::{MinecraftServer, MinecraftServerStatus},
};

use crate::agent::AgentConfig;

use self::{
    config_map::ConfigMapBuilder,
    gameserver::{GameServerBuilder, GameServerBuilderContext},
};

use super::{cluster_ref::resolve_cluster_ref, ReconcilerError, Result};

pub mod config_map;
pub mod gameserver;

#[cfg(test)]
mod fixtures;

struct MinecraftServerReconciler {
    client: kube::Client,
    agent_config: AgentConfig,

    // Builders
    config_map_builder: ConfigMapBuilder,
    gameserver_builder: GameServerBuilder,
}

impl MinecraftServerReconciler {
    async fn reconcile(
        &self,
        api: Api<MinecraftServer>,
        minecraft_server: Arc<MinecraftServer>,
    ) -> Result<Action> {
        let cluster = resolve_cluster_ref(
            &self.client,
            &minecraft_server.namespace().unwrap(),
            &minecraft_server.spec.cluster_ref,
        )
        .await?;

        reconcile_builder(&self.config_map_builder, minecraft_server.as_ref(), None)
            .await
            .map_err(ReconcilerError::BuilderError)?;
        let gameserver = reconcile_builder(
            &self.gameserver_builder,
            minecraft_server.as_ref(),
            Some(GameServerBuilderContext {
                cluster: &cluster,
                agent_config: &self.agent_config,
            }),
        )
        .await
        .map_err(ReconcilerError::BuilderError)?;

        if let Some(gameserver) = &gameserver {
            if let Some(gameserver_status) = &gameserver.status {
                if gameserver_status.state == "Shutdown" {
                    api.delete(&minecraft_server.name_any(), &DeleteParams::default())
                        .await
                        .map_err(ReconcilerError::FailedToDeleteStale)?;

                    return Ok(Action::await_change());
                }

                let mut minecraft_server = minecraft_server.as_ref().clone();
                if minecraft_server.status.is_none() {
                    minecraft_server.status = Some(MinecraftServerStatus::default());
                }

                let status = minecraft_server.status.as_mut().unwrap();

                status.address = gameserver_status.address.clone();
                status.port = gameserver_status.ports.first().unwrap().port;

                if gameserver_status.state == "Ready" || gameserver_status.state == "Allocated" {
                    status.set_condition(
                        "Ready".to_string(),
                        ConditionStatus::True,
                        "ReadyOrAllocated".to_string(),
                        "Server is ready and maybe already allocated".to_string(),
                    );
                } else {
                    status.set_condition(
                        "Ready".to_string(),
                        ConditionStatus::False,
                        "Unknown".to_string(),
                        "Server is not ready yet".to_string(),
                    );
                };

                patch_status(
                    &api,
                    &PatchParams::apply("shulker-operator").force(),
                    &minecraft_server,
                )
                .await
                .map_err(ReconcilerError::BuilderError)?;
            }
        }

        Ok(Action::requeue(Duration::from_secs(5 * 60)))
    }

    async fn cleanup(&self, minecraft_server: Arc<MinecraftServer>) -> Result<Action> {
        info!(
            name = minecraft_server.name_any(),
            namespace = minecraft_server.namespace(),
            "cleaning up MinecraftServer",
        );

        Ok(Action::await_change())
    }

    fn get_labels(
        minecraft_server: &MinecraftServer,
        name: String,
        component: String,
    ) -> BTreeMap<String, String> {
        BTreeMap::from([
            ("app.kubernetes.io/name".to_string(), name.to_string()),
            (
                "app.kubernetes.io/instance".to_string(),
                format!("{}-{}", name, minecraft_server.name_any()),
            ),
            ("app.kubernetes.io/component".to_string(), component),
            (
                "app.kubernetes.io/part-of".to_string(),
                format!("cluster-{}", minecraft_server.spec.cluster_ref.name.clone()),
            ),
            (
                "app.kubernetes.io/managed-by".to_string(),
                "shulker-operator".to_string(),
            ),
            (
                "minecraftcluster.shulkermc.io/name".to_string(),
                minecraft_server.spec.cluster_ref.name.clone(),
            ),
        ])
    }
}

#[instrument(skip(ctx, minecraft_server))]
async fn reconcile(
    minecraft_server: Arc<MinecraftServer>,
    ctx: Arc<MinecraftServerReconciler>,
) -> Result<Action> {
    let ns = minecraft_server.namespace().unwrap();
    let minecraft_servers_api: Api<MinecraftServer> = Api::namespaced(ctx.client.clone(), &ns);

    info!(
        name = minecraft_server.name_any(),
        namespace = ns,
        "reconciling MinecraftServer",
    );

    if minecraft_server.metadata.deletion_timestamp.is_none() {
        ctx.reconcile(minecraft_servers_api.clone(), minecraft_server.clone())
            .await
    } else {
        ctx.cleanup(minecraft_server.clone()).await
    }
}

fn error_policy(
    _minecraft_server: Arc<MinecraftServer>,
    error: &ReconcilerError,
    _ctx: Arc<MinecraftServerReconciler>,
) -> Action {
    warn!("reconcile failed: {:?}", error);
    // ctx.metrics.reconcile_failure(&minecraft_server, error);
    Action::requeue(Duration::from_secs(5))
}

pub async fn run(client: Client, agent_config: AgentConfig) {
    let minecraft_servers_api = Api::<MinecraftServer>::all(client.clone());
    if let Err(e) = minecraft_servers_api
        .list(&ListParams::default().limit(1))
        .await
    {
        error!("CRD is not queryable; {e:?}. Is the CRD installed?");
        std::process::exit(1);
    }

    let context = MinecraftServerReconciler {
        client: client.clone(),
        agent_config,
        config_map_builder: ConfigMapBuilder::new(client.clone()),
        gameserver_builder: GameServerBuilder::new(client.clone()),
    };

    Controller::new(minecraft_servers_api, Config::default().any_semantic())
        .owns(
            Api::<ConfigMap>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .owns(
            Api::<GameServer>::all(client.clone()),
            Config::default().any_semantic(),
        )
        .shutdown_on_signal()
        .run(reconcile, error_policy, context.into())
        .filter_map(|x| async move { std::result::Result::ok(x) })
        .for_each(|_| futures::future::ready(()))
        .await;
}
