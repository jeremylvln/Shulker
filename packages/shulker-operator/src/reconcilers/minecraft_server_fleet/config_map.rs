use k8s_openapi::api::core::v1::ConfigMap;
use kube::core::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use shulker_crds::v1alpha1::minecraft_server_fleet::MinecraftServerFleet;

use super::MinecraftServerFleetReconciler;

pub struct ConfigMapBuilder {
    client: Client,
}

#[async_trait::async_trait]
impl ResourceBuilder for ConfigMapBuilder {
    type OwnerType = MinecraftServerFleet;
    type ResourceType = ConfigMap;

    fn name(minecraft_server_fleet: &Self::OwnerType) -> String {
        format!("{}-config", minecraft_server_fleet.name_any())
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server_fleet: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server_fleet.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server_fleet: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let config_map = ConfigMap {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server_fleet.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerFleetReconciler::get_common_labels(minecraft_server_fleet)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            ..ConfigMap::default()
        };

        Ok(config_map)
    }

    async fn update(
        &self,
        minecraft_server_fleet: &Self::OwnerType,
        config_map: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        config_map.data = Some(
            crate::reconcilers::minecraft_server::config_map::ConfigMapBuilder::get_data_from_spec(
                &minecraft_server_fleet.spec.template.spec.config,
            ),
        );
        Ok(())
    }
}

impl ConfigMapBuilder {
    pub fn new(client: Client) -> Self {
        ConfigMapBuilder { client }
    }
}
