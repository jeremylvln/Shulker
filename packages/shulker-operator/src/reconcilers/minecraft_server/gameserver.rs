use std::collections::BTreeMap;

use k8s_openapi::api::core::v1::Capabilities;
use k8s_openapi::api::core::v1::ConfigMapVolumeSource;
use k8s_openapi::api::core::v1::Container;
use k8s_openapi::api::core::v1::EmptyDirVolumeSource;
use k8s_openapi::api::core::v1::EnvVar;
use k8s_openapi::api::core::v1::EnvVarSource;
use k8s_openapi::api::core::v1::ObjectFieldSelector;
use k8s_openapi::api::core::v1::PodSpec;
use k8s_openapi::api::core::v1::PodTemplateSpec;
use k8s_openapi::api::core::v1::SecretKeySelector;
use k8s_openapi::api::core::v1::SecurityContext;
use k8s_openapi::api::core::v1::Volume;
use k8s_openapi::api::core::v1::VolumeMount;
use k8s_openapi::apimachinery::pkg::apis::meta::v1::ObjectMeta;
use kube::Api;
use kube::Client;
use kube::ResourceExt;

use crate::reconcilers::builder::ResourceBuilder;
use crate::resources::resourceref_resolver::ResourceRefResolver;
use google_agones_crds::v1::game_server::GameServer;
use google_agones_crds::v1::game_server::GameServerEvictionSpec;
use google_agones_crds::v1::game_server::GameServerHealthSpec;
use google_agones_crds::v1::game_server::GameServerPortSpec;
use google_agones_crds::v1::game_server::GameServerSpec;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServer;
use shulker_crds::v1alpha1::minecraft_server::MinecraftServerSpec;

use super::config_map::ConfigMapBuilder;
use super::MinecraftServerReconciler;

const MINECRAFT_SERVER_SHULKER_CONFIG_DIR: &str = "/mnt/shulker/config";
const MINECRAFT_SERVER_CONFIG_DIR: &str = "/config";
const MINECRAFT_SERVER_DATA_DIR: &str = "/data";
const MINECRAFT_SERVER_SHULKER_PROXY_AGENT_VERSION: &str = env!("CARGO_PKG_VERSION");
const MINECRAFT_SERVER_SHULKER_MAVEN_REPOSITORY: &str =
    "https://maven.jeremylvln.fr/artifactory/shulker";

lazy_static! {
    static ref PROXY_SECURITY_CONTEXT: SecurityContext = SecurityContext {
        allow_privilege_escalation: Some(false),
        read_only_root_filesystem: Some(true),
        run_as_non_root: Some(true),
        run_as_user: Some(1000),
        capabilities: Some(Capabilities {
            drop: Some(vec!["ALL".to_string()]),
            ..Capabilities::default()
        }),
        ..SecurityContext::default()
    };
}

pub struct GameServerBuilder {
    client: Client,
    resourceref_resolver: ResourceRefResolver,
}

#[async_trait::async_trait]
impl ResourceBuilder for GameServerBuilder {
    type OwnerType = MinecraftServer;
    type ResourceType = GameServer;

    fn name(minecraft_server: &Self::OwnerType) -> String {
        minecraft_server.name_any()
    }

    fn is_updatable() -> bool {
        true
    }

    fn api(&self, minecraft_server: &Self::OwnerType) -> kube::Api<Self::ResourceType> {
        Api::namespaced(
            self.client.clone(),
            minecraft_server.namespace().as_ref().unwrap(),
        )
    }

    fn is_needed(&self, _minecraft_server: &Self::OwnerType) -> bool {
        true
    }

    async fn create(
        &self,
        minecraft_server: &Self::OwnerType,
        name: &str,
    ) -> Result<Self::ResourceType, anyhow::Error> {
        let game_server = GameServer {
            metadata: ObjectMeta {
                name: Some(name.to_string()),
                namespace: Some(minecraft_server.namespace().unwrap().clone()),
                labels: Some(
                    MinecraftServerReconciler::get_common_labels(minecraft_server)
                        .into_iter()
                        .collect(),
                ),
                ..ObjectMeta::default()
            },
            spec: GameServerSpec::default(),
            status: None,
        };

        Ok(game_server)
    }

    async fn update(
        &self,
        minecraft_server: &Self::OwnerType,
        game_server: &mut Self::ResourceType,
    ) -> Result<(), anyhow::Error> {
        game_server.spec =
            Self::get_game_server_spec(&self.resourceref_resolver, minecraft_server).await?;
        Ok(())
    }
}

impl GameServerBuilder {
    pub fn new(client: Client) -> Self {
        GameServerBuilder {
            client: client.clone(),
            resourceref_resolver: ResourceRefResolver::new(client.clone()),
        }
    }

    pub async fn get_game_server_spec(
        resourceref_resolver: &ResourceRefResolver,
        minecraft_server: &MinecraftServer,
    ) -> Result<GameServerSpec, anyhow::Error> {
        let pod_template_spec =
            Self::get_pod_template_spec(resourceref_resolver, minecraft_server).await?;

        let game_server_spec = GameServerSpec {
            ports: Some(vec![GameServerPortSpec {
                name: "minecraft".to_string(),
                container_port: 25565,
                protocol: "TCP".to_string(),
            }]),
            eviction: Some(GameServerEvictionSpec {
                safe: "OnUpgrade".to_string(),
            }),
            health: Some(GameServerHealthSpec {
                disabled: Some(false),
                initial_delay_seconds: Some(30),
                period_seconds: Some(15),
                failure_threshold: Some(5),
            }),
            template: pod_template_spec,
        };

        Ok(game_server_spec)
    }

    async fn get_pod_template_spec(
        resourceref_resolver: &ResourceRefResolver,
        minecraft_server: &MinecraftServer,
    ) -> Result<PodTemplateSpec, anyhow::Error> {
        let mut pod_spec = PodSpec {
            init_containers: Some(vec![Container {
                image: Some("alpine:latest".to_string()),
                name: "init-fs".to_string(),
                command: Some(vec![
                    "sh".to_string(),
                    format!("{}/init-fs.sh", MINECRAFT_SERVER_SHULKER_CONFIG_DIR),
                ]),
                env: Some(Self::get_init_env(resourceref_resolver, &minecraft_server.spec).await?),
                security_context: Some(PROXY_SECURITY_CONTEXT.clone()),
                volume_mounts: Some(vec![
                    VolumeMount {
                        name: "shulker-config".to_string(),
                        mount_path: MINECRAFT_SERVER_SHULKER_CONFIG_DIR.to_string(),
                        read_only: Some(true),
                        ..VolumeMount::default()
                    },
                    VolumeMount {
                        name: "server-config".to_string(),
                        mount_path: MINECRAFT_SERVER_CONFIG_DIR.to_string(),
                        ..VolumeMount::default()
                    },
                ]),
                ..Container::default()
            }]),
            containers: vec![Container {
                image: Some("itzg/minecraft-server:2022.16.0-java17".to_string()),
                name: "minecraft-server".to_string(),
                env: Some(Self::get_env(&minecraft_server.spec)),
                image_pull_policy: Some("IfNotPresent".to_string()),
                security_context: Some(PROXY_SECURITY_CONTEXT.clone()),
                volume_mounts: Some(vec![
                    VolumeMount {
                        name: "server-config".to_string(),
                        mount_path: MINECRAFT_SERVER_CONFIG_DIR.to_string(),
                        ..VolumeMount::default()
                    },
                    VolumeMount {
                        name: "server-data".to_string(),
                        mount_path: MINECRAFT_SERVER_DATA_DIR.to_string(),
                        ..VolumeMount::default()
                    },
                    VolumeMount {
                        name: "server-tmp".to_string(),
                        mount_path: "/tmp".to_string(),
                        ..VolumeMount::default()
                    },
                ]),
                ..Container::default()
            }],
            service_account_name: Some(format!(
                "{}-server",
                &minecraft_server.spec.cluster_ref.name
            )),
            restart_policy: Some("Never".to_string()),
            volumes: Some(vec![
                Volume {
                    name: "shulker-config".to_string(),
                    config_map: Some(ConfigMapVolumeSource {
                        name: Some(ConfigMapBuilder::name(minecraft_server)),
                        ..ConfigMapVolumeSource::default()
                    }),
                    ..Volume::default()
                },
                Volume {
                    name: "server-config".to_string(),
                    empty_dir: Some(EmptyDirVolumeSource::default()),
                    ..Volume::default()
                },
                Volume {
                    name: "server-data".to_string(),
                    empty_dir: Some(EmptyDirVolumeSource::default()),
                    ..Volume::default()
                },
                Volume {
                    name: "server-tmp".to_string(),
                    empty_dir: Some(EmptyDirVolumeSource::default()),
                    ..Volume::default()
                },
            ]),
            ..PodSpec::default()
        };

        if let Some(pod_overrides) = &minecraft_server.spec.pod_overrides {
            if let Some(image_overrides) = &pod_overrides.image {
                pod_spec.containers[0].image = Some(image_overrides.name.clone());
                pod_spec.containers[0].image_pull_policy =
                    Some(image_overrides.pull_policy.clone());

                if let Some(image_pull_secrets) = pod_spec.image_pull_secrets.as_mut() {
                    image_pull_secrets.append(&mut image_overrides.image_pull_secrets.clone());
                }
            }

            if let Some(resources_overrides) = &pod_overrides.resources {
                pod_spec.containers[0].resources = Some(resources_overrides.clone());
            }

            if let Some(affinity_overrides) = &pod_overrides.affinity {
                pod_spec.affinity = Some(affinity_overrides.clone());
            }

            if let Some(node_selector_overrides) = &pod_overrides.node_selector {
                pod_spec.node_selector =
                    Some(node_selector_overrides.clone().into_iter().collect());
            }

            pod_spec.tolerations = pod_overrides.tolerations.clone();
        }

        Ok(PodTemplateSpec {
            metadata: Some(ObjectMeta {
                labels: Some(MinecraftServerReconciler::get_common_labels(
                    minecraft_server,
                )),
                annotations: Some(BTreeMap::<String, String>::from([(
                    "minecraftserver.shulkermc.io/tags".to_string(),
                    minecraft_server.spec.tags.join(","),
                )])),
                ..ObjectMeta::default()
            }),
            spec: Some(pod_spec),
        })
    }

    async fn get_init_env(
        resourceref_resolver: &ResourceRefResolver,
        spec: &MinecraftServerSpec,
    ) -> Result<Vec<EnvVar>, anyhow::Error> {
        let mut env: Vec<EnvVar> = vec![
            EnvVar {
                name: "SHULKER_CONFIG_DIR".to_string(),
                value: Some(MINECRAFT_SERVER_SHULKER_CONFIG_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SERVER_CONFIG_DIR".to_string(),
                value: Some(MINECRAFT_SERVER_CONFIG_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SERVER_DATA_DIR".to_string(),
                value: Some(MINECRAFT_SERVER_DATA_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "TYPE".to_string(),
                value: Some(Self::get_type_from_version_channel(&spec.version.channel)),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SHULKER_SERVER_AGENT_VERSION".to_string(),
                value: Some(MINECRAFT_SERVER_SHULKER_PROXY_AGENT_VERSION.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SHULKER_MAVEN_REPOSITORY".to_string(),
                value: Some(MINECRAFT_SERVER_SHULKER_MAVEN_REPOSITORY.to_string()),
                ..EnvVar::default()
            },
        ];

        if let Some(world) = &spec.config.world {
            let url = resourceref_resolver.resolve(world).await?.as_url()?;

            env.push(EnvVar {
                name: "SERVER_WORLD_URL".to_string(),
                value: Some(url.to_string()),
                ..EnvVar::default()
            })
        }

        if let Some(plugins) = &spec.config.plugins {
            let urls: Vec<String> = resourceref_resolver
                .resolve_all(plugins)
                .await?
                .into_iter()
                .map(|url| url.to_string())
                .collect();

            env.push(EnvVar {
                name: "SERVER_PLUGIN_URLS".to_string(),
                value: Some(urls.join(",")),
                ..EnvVar::default()
            })
        }

        if let Some(patches) = &spec.config.patches {
            let urls: Vec<String> = resourceref_resolver
                .resolve_all(patches)
                .await?
                .into_iter()
                .map(|url| url.to_string())
                .collect();

            env.push(EnvVar {
                name: "SERVER_PATCH_URLS".to_string(),
                value: Some(urls.join(",")),
                ..EnvVar::default()
            })
        }

        Ok(env)
    }

    fn get_env(spec: &MinecraftServerSpec) -> Vec<EnvVar> {
        let mut env: Vec<EnvVar> = vec![
            EnvVar {
                name: "SHULKER_SERVER_NAME".to_string(),
                value_from: Some(EnvVarSource {
                    field_ref: Some(ObjectFieldSelector {
                        field_path: "metadata.name".to_string(),
                        ..ObjectFieldSelector::default()
                    }),
                    ..EnvVarSource::default()
                }),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SHULKER_SERVER_NAMESPACE".to_string(),
                value_from: Some(EnvVarSource {
                    field_ref: Some(ObjectFieldSelector {
                        field_path: "metadata.namespace".to_string(),
                        ..ObjectFieldSelector::default()
                    }),
                    ..EnvVarSource::default()
                }),
                ..EnvVar::default()
            },
            EnvVar {
                name: "TYPE".to_string(),
                value: Some(Self::get_type_from_version_channel(&spec.version.channel)),
                ..EnvVar::default()
            },
            EnvVar {
                name: Self::get_version_env_from_version_channel(&spec.version.channel),
                value: Some(spec.version.name.clone()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "EULA".to_string(),
                value: Some("TRUE".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "COPY_CONFIG_DEST".to_string(),
                value: Some(MINECRAFT_SERVER_DATA_DIR.to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SYNC_SKIP_NEWER_IN_DESTINATION".to_string(),
                value: Some("true".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "SKIP_SERVER_PROPERTIES".to_string(),
                value: Some("true".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "REPLACE_ENV_IN_PLACE".to_string(),
                value: Some("true".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "REPLACE_ENV_VARIABLE_PREFIX".to_string(),
                value: Some("CFG_".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "CFG_VELOCITY_FORWARDING_SECRET".to_string(),
                value_from: Some(EnvVarSource {
                    secret_key_ref: Some(SecretKeySelector {
                        name: Some(format!("{}-forwarding-secret", spec.cluster_ref.name)),
                        ..SecretKeySelector::default()
                    }),
                    ..EnvVarSource::default()
                }),
                ..EnvVar::default()
            },
            EnvVar {
                name: "MEMORY".to_string(),
                value: Some("".to_string()),
                ..EnvVar::default()
            },
            EnvVar {
                name: "JVM_XX_OPTS".to_string(),
                value: Some("-XX:MaxRAMPercentage=75".to_string()),
                ..EnvVar::default()
            },
        ];

        if let Some(pod_overrides) = &spec.pod_overrides {
            if let Some(env_overrides) = &pod_overrides.env {
                env.extend(env_overrides.clone());
            }
        }

        env
    }

    fn get_type_from_version_channel(channel: &str) -> String {
        match channel {
            "Bukkit" => "BUKKIT".to_string(),
            "Spigot" => "SPIGOT".to_string(),
            "Paper" => "PAPER".to_string(),
            _ => unreachable!("Unknown version channel"),
        }
    }

    fn get_version_env_from_version_channel(_channel: &str) -> String {
        "VERSION".to_string()
    }
}
