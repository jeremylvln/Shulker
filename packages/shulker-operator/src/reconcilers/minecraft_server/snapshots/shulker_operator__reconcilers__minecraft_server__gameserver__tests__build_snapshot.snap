---
source: packages/shulker-operator/src/reconcilers/minecraft_server/gameserver.rs
expression: game_server
---
apiVersion: agones.dev/v1
kind: GameServer
metadata:
  annotations:
    minecraftserver.shulkermc.io/tags: lobby
  labels:
    app.kubernetes.io/component: minecraft-server
    app.kubernetes.io/instance: minecraft-server-my-server
    app.kubernetes.io/managed-by: shulker-operator
    app.kubernetes.io/name: minecraft-server
    app.kubernetes.io/part-of: cluster-my-cluster
    minecraftcluster.shulkermc.io/name: my-cluster
  name: my-server
  namespace: default
spec:
  ports: []
  health:
    disabled: false
    periodSeconds: 15
    failureThreshold: 5
    initialDelaySeconds: 30
  template:
    metadata:
      annotations:
        kubectl.kubernetes.io/default-container: minecraft-server
      labels:
        app.kubernetes.io/component: minecraft-server
        app.kubernetes.io/instance: minecraft-server-my-server
        app.kubernetes.io/managed-by: shulker-operator
        app.kubernetes.io/name: minecraft-server
        app.kubernetes.io/part-of: cluster-my-cluster
        minecraftcluster.shulkermc.io/name: my-cluster
    spec:
      containers:
        - env:
            - name: SHULKER_SERVER_NAME
              valueFrom:
                fieldRef:
                  fieldPath: metadata.name
            - name: SHULKER_SERVER_NAMESPACE
              valueFrom:
                fieldRef:
                  fieldPath: metadata.namespace
            - name: SHULKER_NETWORK_ADMINS
              value: ""
            - name: SHULKER_SERVER_LIFECYCLE_STRATEGY
              value: AllocateWhenNotEmpty
            - name: EULA
              value: "TRUE"
            - name: COPY_CONFIG_DEST
              value: /data
            - name: SYNC_SKIP_NEWER_IN_DESTINATION
              value: "true"
            - name: SKIP_SERVER_PROPERTIES
              value: "true"
            - name: REPLACE_ENV_IN_PLACE
              value: "true"
            - name: REPLACE_ENV_VARIABLE_PREFIX
              value: CFG_
            - name: CFG_VELOCITY_FORWARDING_SECRET
              valueFrom:
                secretKeyRef:
                  key: key
                  name: my-cluster-forwarding-secret
            - name: MEMORY
              value: ""
            - name: JVM_XX_OPTS
              value: "-XX:MaxRAMPercentage=75"
            - name: TYPE
              value: PAPER
            - name: VERSION
              value: 1.20.1
            - name: EXTRA_ENV
              value: my_value
          image: "itzg/minecraft-server:2025.1.0-java21"
          imagePullPolicy: IfNotPresent
          name: minecraft-server
          ports:
            - containerPort: 25565
              name: minecraft
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
                - ALL
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 1000
          volumeMounts:
            - mountPath: /config
              name: server-config
            - mountPath: /data
              name: server-data
            - mountPath: /tmp
              name: server-tmp
      initContainers:
        - command:
            - sh
            - /mnt/shulker/config/init-fs.sh
          env:
            - name: SHULKER_CONFIG_DIR
              value: /mnt/shulker/config
            - name: SHULKER_SERVER_CONFIG_DIR
              value: /config
            - name: SHULKER_SERVER_DATA_DIR
              value: /data
            - name: SHULKER_VERSION_CHANNEL
              value: Paper
            - name: SHULKER_SERVER_WORLD_URL
              value: "https://example.com/my_world.tar.gz"
            - name: SHULKER_SERVER_PLUGIN_URLS
              value: "https://maven.jeremylvln.fr/repository/shulker-snapshots/io/shulkermc/shulker-server-agent/0.0.0-test-cfg/shulker-server-agent-0.0.0-test-cfg-paper.jar;https://example.com/my_plugin.jar"
            - name: SHULKER_SERVER_PATCH_URLS
              value: "https://example.com/my_patch.tar.gz"
          image: "alpine:latest"
          name: init-fs
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
                - ALL
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 1000
          volumeMounts:
            - mountPath: /mnt/shulker/config
              name: shulker-config
              readOnly: true
            - mountPath: /config
              name: server-config
      nodeSelector:
        beta.kubernetes.io/os: linux
      restartPolicy: Never
      serviceAccountName: shulker-my-cluster-server
      subdomain: my-cluster-cluster
      volumes:
        - configMap:
            name: my-server-config
          name: shulker-config
        - emptyDir: {}
          name: server-config
        - emptyDir: {}
          name: server-data
        - emptyDir: {}
          name: server-tmp
  eviction:
    safe: OnUpgrade
